spectator-api/src/main/java/com/netflix/spectator/api/patterns/CardinalityLimiters.java

```diff
@@ -357,4 +379,39 @@ private static class RollupLimiter implements Function<String, String>, Serializ
       return "RollupLimiter(" + state + ")";
     }
   }
+
+  private static class RegisteredNameOrIpLimiter implements Function<String, String>, Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    //From RFC 3986 3.2.2, we can quickly identify IP literals (other than IPv4) from the first character.
+    private static final Predicate<String> IS_IP_LITERAL = s -> s.startsWith("[");
+
+    //From RFC 3986 3.2.2
+    private static final String DEC_OCTET_PATTERN = "(?:\\d|[1-9]\\d|1\\d{2}|2[0-4]\\d|25[0-5])";
```

Comment:
Multiple OR clauses will be significantly slower to match than just the simple `\d{1-3}` check used before. Not sure we really care about strict spec compliance for this in particular since IPv6 is just checking that it starts with a `[`.

Reply:
I adjusted the implementation to align with the spec mentioned in the Javadoc. I'm happy to revert to the simpler expression and do some more wordsmithing in the docs.

Reply:
I guess to me what is odd about it is the discrepancy between the IPv4 and IPv6 checks. One is trying to be really precise and strictly match based on the spec and the other is just a simple cheap check that indicates it is likely to be an IPv6 address. If we are going to strictly check it is a valid IPv4 address according to the spec, then I would argue the same should be done for IPv6.

For this particular use-case, I think a simple check that something is likely to be an IP address is good enough and in most cases the performance benefit would be preferable. For example, for the purposes of the limiter is there harm to flagging 987.0.0.1 as an IP even though strictly speaking the number is out of range?

Reply:
My intent was to leverage the spec to write lower-cost checks. A `reg-name` cannot include a "[", so assuming that the input is a 3.2.2 `host`, it suffices to check the first character. Per the spec, an out-of-range IPv4 address would be a reg-name, hence the stricter check.

Reply:
I'm less confident it will follow that spec. In particular for Eureka VIPs which if I understand correctly are just arbitrary strings. Setup a simple benchmark to compare the more precise regex to the simpler variant:

```
Benchmark                         Mode  Cnt          Score         Error   Units
hostPrecise                      thrpt    5   21563373.607 ±  264239.452   ops/s
hostSimple                       thrpt    5  261748216.141 ± 2551757.877   ops/s
ipPrecise                        thrpt    5    6223508.840 ±    9284.224   ops/s
ipSimple                         thrpt    5   16061395.232 ±   28172.942   ops/s
```

It was about 10x slower for what I would expect to be the most common case of a host.

Reply:
Thanks for taking the time to run some benchmarks. The current version of the code uses the `\d{1,3}\.` form of the matcher.

You are correct that Eureka VIPs are arbitrary strings. This has been sufficiently problematic that we've taken some steps to clean them up.